### PR TITLE
Query-frontend: fix silent panic in remote read API if there's no matcher

### PR DESCRIPTION
**Pull request by pracucci**
##### The problem

[This PR ](https://github.com/grafana/mimir/pull/12969) introduced a regression in the remote read API, causing the following silent panic catched by HTTP handler:
```
http: panic serving IP:PORT: promql.ChildrenIter: unhandled node type <nil>
```

The key change introduced by the PR is that previously the query-frontend middlewares were directly calling `parser.ParseExpr()` to clone a query, and this returns an error if the input expression is an empty string (returns `unknown position: parse error: no expression found in input` error).

After the PR change, if the request has no query expression, `req.GetParsedQuery()` returns `nil`. Then, middlewares call `astmapper.CloneExpr(nil)` that simply returns `nil`, and so the middleware execution continues, triggering the panic.

##### The proposed solution

At first I've considered checking for `nil` returned by `req.GetParsedQuery()` in every place where we call it. However, I think it may be too fragile: in the future we may easily forget to check for `nil`, re-introducing a similar regression.

So, in this PR I'm proposing a different approach:

1. Rename `req.GetParsedQuery()` to `req.GetClonedParsedQuery()` and always return a clone (we were cloning in each caller anyway)
2. Return an error from `req.GetClonedParsedQuery()` if there's no query. This makes it more loud to the caller.

##### Notes about tests

The new test case introduced in `TestTripperware_RemoteRead` panics in the `main` branch:

```
--- FAIL: TestTripperware_RemoteRead (0.00s)
panic: promql.ChildrenIter: unhandled node type <nil> [recovered, repanicked]

goroutine 67 [running]:
testing.tRunner.func1.2({0x1063be720, 0x140004fc650})
	/Users/marco/workspace/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.4.darwin-arm64/src/testing/testing.go:1872 +0x190
testing.tRunner.func1()
	/Users/marco/workspace/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.4.darwin-arm64/src/testing/testing.go:1875 +0x31c
panic({0x1063be720?, 0x140004fc650?})
	/Users/marco/workspace/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.4.darwin-arm64/src/runtime/panic.go:783 +0x120
github.com/prometheus/prometheus/promql/parser.Walk.ChildrenIter.func1(...)
	/Users/marco/workspace/src/github.com/grafana/mimir/vendor/github.com/prometheus/prometheus/promql/parser/ast.go:434
github.com/prometheus/prometheus/promql/parser.Walk({0x1068de1a0?, 0x140004fc640?}, {0x0, 0x0}, {0x140006f4280, 0x0, 0x4})
	/Users/marco/workspace/src/github.com/grafana/mimir/vendor/github.com/prometheus/prometheus/promql/parser/ast.go:344 +0x660
github.com/prometheus/prometheus/promql/parser.Inspect(...)
	/Users/marco/workspace/src/github.com/grafana/mimir/vendor/github.com/prometheus/prometheus/promql/parser/ast.go:385
github.com/prometheus/prometheus/promql/parser.ExtractSelectors({0x0, 0x0})
	/Users/marco/workspace/src/github.com/grafana/mimir/vendor/github.com/prometheus/prometheus/promql/parser/ast.go:359 +0xc8
github.com/grafana/mimir/pkg/frontend/querymiddleware.queryStatsMiddleware.trackRegexpMatchers({0x45d964b800, {0x10691d828, 0x140004f63c0}, {0x10691d828, 0x140004f6480}, 0x140003bc970, 0x140003bc978, {0x1068dd900, 0x140005f62d0}}, {0x10693b350?, ...})
	/Users/marco/workspace/src/github.com/grafana/mimir/pkg/frontend/querymiddleware/stats.go:79 +0x4c
github.com/grafana/mimir/pkg/frontend/querymiddleware.queryStatsMiddleware.Do({0x45d964b800, {0x10691d828, 0x140004f63c0}, {0x10691d828, 0x140004f6480}, 0x140003bc970, 0x140003bc978, {0x1068dd900, 0x140005f62d0}}, {0x10690d4c8, ...}, ...)
	/Users/marco/workspace/src/github.com/grafana/mimir/pkg/frontend/querymiddleware/stats.go:66 +0x58
github.com/grafana/mimir/pkg/frontend/querymiddleware.(*remoteReadRoundTripper).RoundTrip(0x140003a9a80, 0x140002cef00)
	/Users/marco/workspace/src/github.com/grafana/mimir/pkg/frontend/querymiddleware/remote_read.go:86 +0x1a4
github.com/grafana/mimir/pkg/frontend/querymiddleware.newQueryTripperware.func1.1(0x140002cef00)
	/Users/marco/workspace/src/github.com/grafana/mimir/pkg/frontend/querymiddleware/roundtrip.go:391 +0x2d0
github.com/grafana/mimir/pkg/frontend/querymiddleware.RoundTripFunc.RoundTrip(0x140004fc580?, 0x140006f38e8?)
	/Users/marco/workspace/src/github.com/grafana/mimir/pkg/frontend/querymiddleware/roundtrip.go:221 +0x28
github.com/grafana/mimir/pkg/frontend/querymiddleware.newQueryCountTripperware.func2.1(0x140002cef00)
	/Users/marco/workspace/src/github.com/grafana/mimir/pkg/frontend/querymiddleware/roundtrip.go:650 +0x3e0
github.com/grafana/mimir/pkg/frontend/querymiddleware.RoundTripFunc.RoundTrip(0x1068dd2a0?, 0x140003a9840?)
	/Users/marco/workspace/src/github.com/grafana/mimir/pkg/frontend/querymiddleware/roundtrip.go:221 +0x28
github.com/grafana/mimir/pkg/frontend/querymiddleware.TestTripperware_RemoteRead.func5(0x14000103880)
	/Users/marco/workspace/src/github.com/grafana/mimir/pkg/frontend/querymiddleware/roundtrip_test.go:852 +0x458
testing.tRunner(0x14000103880, 0x140007766f0)
	/Users/marco/workspace/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.4.darwin-arm64/src/testing/testing.go:1934 +0xc8
created by testing.(*T).Run in goroutine 33
	/Users/marco/workspace/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.4.darwin-arm64/src/testing/testing.go:1997 +0x364

```

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`. If changelog entry is not needed, please add the `changelog-not-needed` label to the PR.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

---

> [!NOTE]
> Prevents query-frontend panics by introducing GetClonedParsedQuery() with explicit erroring on missing/invalid queries, updating all middlewares and remote read to use it, and adding tests; includes changelog entry and a remote read test for requests without matchers.
> 
> - **API/Interfaces**:
>   - Replace `MetricsQueryRequest.GetParsedQuery()` with `GetClonedParsedQuery()` that clones the AST and returns an error when the request has no query; add `errRequestNoQuery`.
> - **Query-frontend Middlewares**:
>   - Update `blockInternalFunctions`, `durations`, `experimental_functions`, `rewrite`, `querysharding`, `split_and_cache`, and `stats` to use `GetClonedParsedQuery()` and handle errors.
>   - Ensure `spin_off_subqueries` uses the new accessor and returns a proper error on parse failure.
> - **Remote Read**:
>   - Implement `remoteReadQueryRequest.GetClonedParsedQuery()` (parse string, error if empty) to stop silent panics on requests without matchers.
> - **Tests**:
>   - Add non-panic tests for nil/invalid query expressions across middlewares.
>   - Extend roundtrip tests with a remote read request without matchers.
> - **Changelog**:
>   - Add bugfix entry for silent panic on remote read with no matchers.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 88aac9ecaec4149b4ae07cc532cc26646f8ff643. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: pkg/frontend/querymiddleware/codec.go [hunk 1]
```diff
@@ -1336,3 +1341,15 @@ func HeadersToPropagateFromContext(ctx context.Context) map[string][]string {
 
 	return nil
 }
+
+// cloneParsedQuery returns a clone of the input parser.Expr. If an error is provided in input, it returns
+// the provided error as is, without even trying to clone the expression.
+//
+// This utility function is intended to be used to easily clone expressions returned by GetParsedQuery(), e.g.:
+// cloneParsedQuery(req.GetParsedQuer())
+func cloneParsedQuery(expr parser.Expr, err error) (parser.Expr, error) {
```
Comment (56quarters) [inline pkg/frontend/querymiddleware/codec.go]: It looks like we now call this at every callsite for `GetParsedQuery()` except for a single one (the `rewriteMiddleware`) which then does its own clone anyway. Maybe we should just make this part of the implementation and contract of `GetParsedQuery()`?

Comment (pracucci) [reply pkg/frontend/querymiddleware/codec.go]: Good thought. Let me do it.

Comment (pracucci) [reply pkg/frontend/querymiddleware/codec.go]: Done in https://github.com/grafana/mimir/pull/13745/commits/dd7e01e65f661e942672750c7bba57ada8ed4724

File: CHANGELOG.md [hunk 2]
```diff
@@ -101,6 +101,7 @@
 * [BUGFIX] Rename `cortex_bucket_store_cached_postings_compression_time_seconds`, `cortex_query_frontend_regexp_matcher_count`, and `cortex_query_frontend_regexp_matcher_optimized_count` to follow naming conventions. #13599
 * [BUGFIX] Query-frontend: Fix incorrect query results when running sharding inside MQE is enabled and the query contains a subquery eligible for subquery spin-off wrapped in a shardable aggregation. #13619
 * [BUGFIX] Memberlist: Fix occasional nil pointer dereference panics. #13635
+* [BUGFIX] Query-frontend: Fix silence panic when executing a remote read API request if the request has no matchers. #13745
```
Comment (56quarters) [inline CHANGELOG.md]:
```suggestion
* [BUGFIX] Query-frontend: Fix silent panic when executing a remote read API request if the request has no matchers. #13745
```

Comment (gotjosh) [inline CHANGELOG.md]:
```suggestion
* [BUGFIX] Query-frontend: Fix silent panic when executing a remote read API request if the request has no matchers. #13745
```

**Review by gotjosh — APPROVED** [review]
LGTM

With the direction the change went - I'm not sure how much value the unit tests add, I guess it's better than no coverage?

**Comment by pracucci** [comment]
> I'm not sure how much value the unit tests add, I guess it's better than no coverage?

I don't think they hurt.
